### PR TITLE
build: relieve the Makefile of release configurations

### PR DESCRIPTION
**Pull request by benesch**
The Makefile previously hard-coded some "release configurations", that is,
collections of compiler paths and flags for building releases that targeted
various platforms with various features enabled, like MemorySanitizer (MSan)
and static linking.

These release configurations do not belong in Make. The logic was exceptionally
hard to follow, for one. The release configurations also hardcoded paths to
compiler toolchains in the builder image; they were entirely unsable unless
running Make within the builder image.

Instead, Make should provide the tools to apply a release configuration. This
patch exposes a set of variables (XCC, XCMAKE_SYSTEM_NAME, etc.) that control
cross-compilation. A new program, mkrelease, is baked into the builder image
and sets these variables appropriately. It is appropriate for mkrelease to
hardcode paths to compiler toolchains as it is *part* of the builder image.

This improves readability and provides a nice separation of concerns.

### CODE REVIEW

**Comment by tschottdorf** [comment]
LGTM on the approach and the parts that I've looked at LGTM, but someone else should take a look as well.

---

Reviewed 9 of 9 files at r1.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[build/builder/mkrelease.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27115#-LGTXEeMErypSqQs16VX:-LGTXEeMErypSqQs16VY:bb6ht2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/eae59d5215fda83a6e82571905be56dc4f16e8c3/build/builder/mkrelease.go#L1)):*
> ```Go
> // Copyright 2017 The Cockroach Authors.
> ```

2018

---

*[build/builder/mkrelease.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27115#-LGTYYjG-Zr4xfISL_65:-LGTYYjG-Zr4xfISL_66:b-oifqsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/eae59d5215fda83a6e82571905be56dc4f16e8c3/build/builder/mkrelease.go#L144)):*
> ```Go
> 		out = append(out, shellEscape(e))
> 	}
> 	out = append(out, "make")
> ```

s/"make"/cmd/

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27115#-:-LGTYCHk0uMd1wkR8Qt8:bsnlkci)*

**Comment by petermattis** [comment]
:lgtm: modulo my suggestions on `mkrelease.go`.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[build/builder/mkrelease.go, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27115#-LGUyrJb1egLHkKsEwsk:-LGUyrJb1egLHkKsEwsl:b-csjj9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/eae59d5215fda83a6e82571905be56dc4f16e8c3/build/builder/mkrelease.go#L87)):*
> ```Go
> 		triple: "x86_64-w64-mingw32",
> 	},
> }
> ```

I can't believe I'm suggesting this: would this be simpler/easier to understand as a shell script?

Alternately, we could package each of these settings into a separate Makefile and then have that Makefile invoke the `cockroach/Makefile` similar to the way `cockroach/pkg/Makefile` works.  Something like:

```
%: default
        @true

.PHONY: default
default:
        @$(MAKE) -C .. $(MAKECMDGOALS) TARGET_TRIPE=x XCMAKE_SYSTEM_NAME=y ...
```

That said, I'm not opposed to use Go for this purpose, it just feels a bit unusual.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27115#-:-LGUzo_NCKX14c75DQzM:b-otbtoc)*

**Comment by benesch** [comment]
Review status: :shipit: complete! 1 of 0 LGTMs obtained

---

*[build/builder/mkrelease.go, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27115#-LGUyrJb1egLHkKsEwsk:-LGVCJrp404DxNFXgige:b-m1iq24) ([raw file](https://github.com/cockroachdb/cockroach/blob/eae59d5215fda83a6e82571905be56dc4f16e8c3/build/builder/mkrelease.go#L87)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I can't believe I'm suggesting this: would this be simpler/easier to understand as a shell script?

Alternately, we could package each of these settings into a separate Makefile and then have that Makefile invoke the `cockroach/Makefile` similar to the way `cockroach/pkg/Makefile` works.  Something like:

```
%: default
        @true

.PHONY: default
default:
        @$(MAKE) -C .. $(MAKECMDGOALS) TARGET_TRIPE=x XCMAKE_SYSTEM_NAME=y ...
```

That said, I'm not opposed to use Go for this purpose, it just feels a bit unusual.
</blockquote></details>

Ha! This code was originally #15337, where it started life as a shell script but became a Go program. I also think it's easier to understand as a shell script. I'll switch it back.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27115)*

**Comment by benesch** [comment]
Ok, this is ready for another look! I'm expecting CI to fail because I haven't updated the builder image yet.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[build/builder/mkrelease.go, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27115#-LGTXEeMErypSqQs16VX:-LGVdFir2favuKOhvGqo:b-teqwwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/eae59d5215fda83a6e82571905be56dc4f16e8c3/build/builder/mkrelease.go#L1)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

2018
</blockquote></details>

File removed.

---

*[build/builder/mkrelease.go, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27115#-LGUyrJb1egLHkKsEwsk:-LGVdGFN9ecKiYhf4CWM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/eae59d5215fda83a6e82571905be56dc4f16e8c3/build/builder/mkrelease.go#L87)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Ha! This code was originally #15337, where it started life as a shell script but became a Go program. I also think it's easier to understand as a shell script. I'll switch it back.
</blockquote></details>

Done.

---

*[build/builder/mkrelease.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27115#-LGTYYjG-Zr4xfISL_65:-LGVdCTdAJU-X4vAi58j:b-teqwwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/eae59d5215fda83a6e82571905be56dc4f16e8c3/build/builder/mkrelease.go#L144)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

s/"make"/cmd/
</blockquote></details>

File removed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27115#-:-LGVdHrn0e8MEVFH67ni:bsrwwhl)*

**Comment by tschottdorf** [comment]
Reviewed 6 of 6 files at r2.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27115)*

**Review by petermattis — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27115#-:-LGzVkox-K4C5bqLyHzQ:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by benesch** [comment]
bors r=petermattis

**Comment by benesch** [comment]
bors r=petermattis
